lib/mobile_app_backend/global_data_cache.ex

```diff
@@ -21,43 +23,105 @@ defmodule MobileAppBackend.GlobalDataCache do
         }
 
   defmodule State do
-    defstruct [:key, :update_ms]
+    defstruct [:key, :update_ms, :first_update_ms]
   end
 
-  def default_key, do: __MODULE__
+  @callback default_key :: key()
+  @callback get_data(key()) :: data()
 
-  def start_link(opts) do
+  def start_link(opts \\ []) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).start_link(opts)
+  end
+
+  def handle_info(msg, state) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).handle_info(msg, state)
+  end
+
+  def init(opts \\ []) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).init(opts)
+  end
+
+  def default_key do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).default_key()
+  end
+
+  def get_data(key \\ default_key()) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).get_data(key)
+  end
+end
+
+defmodule MobileAppBackend.GlobalDataCache.Impl do
+  use GenServer
+  alias MBTAV3API.{JsonApi, Repository}
+  alias MobileAppBackend.GlobalDataCache
+  alias MobileAppBackend.GlobalDataCache.State
+
+  @behaviour GlobalDataCache
+
+  @impl true
+  def default_key, do: MobileAppBackend.GlobalDataCache
+
+  @spec start_link(keyword()) :: :ignore | {:error, any()} | {:ok, pid()}
+  def start_link(opts \\ []) do
     opts = Keyword.merge([key: default_key()], opts)
-    GenServer.start_link(__MODULE__, opts)
+    GenServer.start_link(GlobalDataCache, opts)
   end
 
-  @spec get_data(key()) :: data()
+  @impl true
   def get_data(key \\ default_key()) do
     :persistent_term.get(key, nil) || update_data(key)
   end
 
   @impl GenServer
-  def init(opts) do
-    opts = Keyword.merge(Application.get_env(:mobile_app_backend, __MODULE__), opts)
+  def init(opts \\ []) do
+    opts = Keyword.merge(Application.get_env(:mobile_app_backend, GlobalDataCache), opts)
+    first_update_ms = opts[:first_update_ms] || :timer.seconds(1)
 
     state = %State{
       key: opts[:key],
       update_ms: opts[:update_ms] || :timer.minutes(5)
     }
 
+    Process.send_after(self(), :recalculate, first_update_ms)
+
     {:ok, state}
   end
 
   @impl GenServer
-  def handle_info(:recalculate, %State{} = state) do
+  def handle_info(:recalculate, %State{update_ms: update_ms} = state) do
     update_data(state.key)
 
-    Process.send_after(self(), :recalculate, state.update_ms)
+    if :persistent_term.get(state.key, nil) do
```

Comment:
I still don't think this `if` can ever fail.

Reply:
Oh right I meant to back that out too, thanks!

Reply:
Oh wait this if can fail - it will fail the first time `:recalculate` is called since `update_data` isn't called from `init`

Reply:
How will that make it fail? `update_data` (which calls `:persistent_term.put/2`) is called right before this `if`.

Reply:
🤦‍♀️ I should have looked 1 line higher